gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java

```diff
@@ -222,6 +223,80 @@ private void preserveFileAttrInPublisher(CopyableFile copyableFile) throws IOExc
     }
   }
 
+  /** Organizes and encapsulates access to {@link WorkUnitState}s according to useful access patterns. */
+  @AllArgsConstructor
+  private static class WorkUnitStatesHelper {
```

Comment:
Maybe not for this PR, but if this stems from multiple categories of workloads being misidentified maybe we should have this as its own separate class and have a stronger type system around workunit collections?

Reply:
I don't see mis-identification as the underlying reason for the regression.  rather it was more likely not appreciating how presumptive `WorkUnitState::getWorkingState()` is about that WU having already executed.  I doubt it was clear (certainly it shocked me!) that it would fall back on the enclosing job's state to provide state for a WU that had never itself started.

as for that `WorkUnitStatesHelper`, I don't presently foresee utility anywhere beyond `CopyDataPublisher`